src/main/groovy/de/dkfz/roddy/tools/EscapableString.groovy

```diff
@@ -0,0 +1,291 @@
+                if (acc == c()) {
+                    c(v)
+                } else if (v == c()) {
+                    acc
+                } else {
+                    acc + separator + v
+                }
+            })
+        }
+        static ConcatenatedString join(List<EscapableString> list, String separator) {
+            join(list, u(separator))
+        }
+
+        static String forBash(EscapableString string) {
+            string.interpreted(BashInterpreter.instance)
+        }
+
+    }
+
+}
+
+@CompileStatic
+@EqualsAndHashCode(includeFields = true)
+class UnescapedString extends EscapableString {
+
+    private String value
+
+    UnescapedString(@NotNull String value) {
+        Preconditions.checkArgument(value != null)
+        this.value = value
+    }
+
+    String getValue() {
+        return value
+    }
+
+    @Override
+    String toString() {
+        "u($value)"
+    }
+
+    @Override
+    String interpreted(EscapableStringInterpreter interpreter) {
+        interpreter.from(this)
+    }
+
+    @Override
+    EscapedString getEscaped() {
+        new EscapedString(this)
+    }
+
+    @Override
+    int size() {
+        value.size()
+    }
+
+}
+
+@CompileStatic
+@EqualsAndHashCode(includeFields = true)
+class EscapedString extends EscapableString {
+
+    private EscapableString value
+
+    // The EscapableStringInterpreter needs access to the value. For now package private access
+    // is sufficient.
+    @PackageScope
+    EscapableString get_value() {
+        value
+    }
+
+    EscapedString(@NotNull EscapableString value) {
+        Preconditions.checkArgument(value != null)
+        this.value = value
+    }
+
+    @Override
+    String toString() {
+        "e($value)"
+    }
+
+    @Override
+    String interpreted(EscapableStringInterpreter interpreter) {
+        interpreter.from(this)
+    }
+
+    @Override
+    EscapedString getEscaped() {
+        // Stack escapes.
+        new EscapedString(this)
+    }
+
+    @Override
+    int size() {
+        value.size()
+    }
+
+}
+
+@CompileStatic
+@EqualsAndHashCode(includeFields = true)
+class ConcatenatedString extends EscapableString {
+
+    private List<EscapableString> values
+
+    @PackageScope
+    List<EscapableString> get_values() {
+        values
+    }
+
+    ConcatenatedString(@NotNull List<EscapableString> values) {
+        Preconditions.checkArgument(values != null)
+        this.values = values
+    }
+
+    @Override
+    ConcatenatedString getEscaped() {
+        // Escape all values.
+        new ConcatenatedString(values.collect { it.escaped })
+    }
+
+    @Override
+    int size() {
+        values.collect { it.size() }.
+                inject(0, { acc, v -> acc + v })
+    }
+
+    @Override
+    String interpreted(EscapableStringInterpreter interpreter) {
+        interpreter.from(this)
+    }
+
+    @Override
+    String toString() {
+        "c($values)"
+    }
+
+    @Override
+    ConcatenatedString simplify() {
```

Comment:
might be good to add a test for this

Reply:
Kind of true, but the method is actually implicitly tested by other tests already, because it is used only internally for simplifying strings representations.

Instead, I make the method protected.